### ISSUE TITLE
Improve general performance

### ISSUE THREAD

**Issue by lorenzhs** [glowing-bear/glowing-bear#87]
The number of objects created is outrageous and should be reduced dramatically. There may well be over 500.000 objects, and subsequently it takes a long time to switch channels if there is a lot of backlog.

[Screenshot](http://4z2.de/profiling.png) of the Chromium heap profiler. Only action was switching channels from #cyanogenmod-dev to #Cyanogenmod on freenode. The CPU profiler says that over 50% of the time are spent getting innerText (angular-sanitize.js:237).

**Comment by leafo** [comment]
Not sure if I should create new issue for this, but I've noticed that switching to buffers blocks the browser main thread for a considerable amount of time: http://leafo.net/shotsnb/2016-04-12_22-18-47.png

Appears to be worse on channels that have buffers full of quit/join messages.

**Comment by lorenzhs** [comment]
Yes, that's an unfortunate issue in buffers with lots of activity. It depends mostly on the number of lines displayed, I think (filtered/hidden lines are discarded right away). There's a lot of angular processing going on, and some browser layouting (tabular layouts are kinda expensive). Plus of course requesting and waiting for lines If you have ideas how to speed that up (or even want to send a PR), we'd greatly appreciate it! I don't know much about front-end development, so I rely on pointers from others to read up on things and try to apply them to GB ;)

**Comment by blaskovic** [comment]
Is there any progress on this? Running glowing-bear on machine with small RAM (2G for example) is pretty hard. Glowing bear tab could eat up to 400M which I consider a lot!
Refreshing from time to time helps a lot. But that's not a solution..

**Comment by lorenzhs** [comment]
Honestly it's not that much of a priority, since it works flawlessly on new-ish smart phones and tablets. My laptop is over five years old and works fine. Targeting low-end machines would be a *lot* of work, and none of the developers seems interested in that. However, if you want to work on this yourself, that would be very welcome—I'd gladly review a PR and help you with any questions.

**Comment by leafo** [comment]
It's definitely still a problem for me. If I have a buffer that has a few thousand unread messages (which isn't uncommon), then the browser will block for a long time. I'm using a mostly modern desktop PC. I just opened a buffer with 10k unread lines and it took over a minute to finally render.

**Comment by Vaesper** [comment]
Hmm that might be fixed in the (soon-to-be-released) dev version, did you try it in that?

**Comment by leafo** [comment]
I'm running the version that's hosted on https://www.glowing-bear.org

BTW, even just switching between buffers with no unreads produces a significant delay, 3 seconds:

![](http://leafo.net/shotsnb/2016-12-19_15-55-34.png)

**Comment by Vaesper** [comment]
Does the problem persist if you use https://latest.glowing-bear.org ?

**Comment by lorenzhs** [comment]
Hmm, for me it's 330ms, of which 300ms are spent in `$digest` (angular), most of which seems to be checking watchers (I think that corresponds to the narrow blue bars in your screenshot). That's certainly not ideal. But honestly if it takes 10x as long for you then something's badly wrong. Do you have any browser plugins that might interfere with things? What does "mostly modern" mean?

For buffers with no or few unread lines, there shouldn't be any performance difference between latest (master) and www (gh-pages).

**Comment by AStove** [comment]
> switching between buffers with no unreads produces a significant delay, 3 seconds. Maybe even worse than last I noted the issue further up in this thread.

I've made an issue about this and tried to improve it: #1100

**Comment by AStove** [comment]
@leafo could you check if this problem only occured when the dev console is open?
